### PR TITLE
Add table aws_dms_replication_instance. closes #298

### LINKED ISSUE #298: Add table aws_database migration service_replication instance
**References**
To accomplish the below  requirement from AWS Foundational Security Best Practices we need this table.
- Database Migration Service replication instances should not be public

### PR DESCRIPTION
**Pull request by sayan133**
# Integration test logs
<details>
  <summary>Logs</summary>
  
```
No env file present for the current environment:  staging 
 Falling back to .env config
No env file present for the current environment:  staging
customEnv TURBOT_TEST_EXPECTED_TIMEOUT undefined

SETUP: tests/aws_dms_replication_instance []

PRETEST: tests/aws_dms_replication_instance

TEST: tests/aws_dms_replication_instance
Running terraform
aws_dms_replication_instance.named_test_resource: Creating...
aws_dms_replication_instance.named_test_resource: Still creating... [10s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [20s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [30s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [40s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [50s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [1m0s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [1m10s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [1m20s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [1m30s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [1m40s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [1m50s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [2m0s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [2m10s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [2m20s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [2m30s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [2m40s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [2m50s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [3m0s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [3m10s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [3m20s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [3m30s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [3m40s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [3m50s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [4m0s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [4m10s elapsed]
aws_dms_replication_instance.named_test_resource: Still creating... [4m20s elapsed]
aws_dms_replication_instance.named_test_resource: Creation complete after 4m22s [id=turbottest17690]

Warning: Deprecated Resource

The null_data_source was historically used to construct intermediate values to
re-use elsewhere in configuration, the same can now be achieved using locals


Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

Outputs:

account_id = "388460667113"
aws_partition = "aws"
aws_region = "us-east-1"
resource_aka = "arn:aws:dms:us-east-1:388460667113:rep:JPJLT3VN3X75TXNK22G27HNKWIE7KE2X74BRAIQ"
resource_name = "turbottest17690"

Running SQL query: test-get-query.sql
[
  {
    "account_id": "388460667113",
    "akas": [
      "arn:aws:dms:us-east-1:388460667113:rep:JPJLT3VN3X75TXNK22G27HNKWIE7KE2X74BRAIQ"
    ],
    "allocated_storage": 5,
    "auto_minor_version_upgrade": true,
    "availability_zone": "us-east-1a",
    "dns_name_servers": null,
    "free_until": null,
    "multi_az": false,
    "partition": "aws",
    "preferred_maintenance_window": "sun:10:30-sun:14:30",
    "publicly_accessible": false,
    "region": "us-east-1",
    "replication_instance_arn": "arn:aws:dms:us-east-1:388460667113:rep:JPJLT3VN3X75TXNK22G27HNKWIE7KE2X74BRAIQ",
    "replication_instance_class": "dms.t2.micro",
    "replication_instance_identifier": "turbottest17690",
    "replication_instance_status": "available",
    "secondary_availability_zone": null,
    "tags": {
      "foo": "bar"
    },
    "tags_src": [
      {
        "Key": "foo",
        "Value": "bar"
      }
    ],
    "title": "turbottest17690"
  }
]
✔ PASSED

Running SQL query: test-hydrate-query.sql
[
  {
    "region": "us-east-1",
    "replication_instance_arn": "arn:aws:dms:us-east-1:388460667113:rep:JPJLT3VN3X75TXNK22G27HNKWIE7KE2X74BRAIQ",
    "replication_instance_identifier": "turbottest17690",
    "tags": {
      "foo": "bar"
    },
    "tags_src": [
      {
        "Key": "foo",
        "Value": "bar"
      }
    ],
    "title": "turbottest17690"
  }
]
✔ PASSED

Running SQL query: test-list-query.sql
[
  {
    "publicly_accessible": false,
    "region": "us-east-1",
    "replication_instance_arn": "arn:aws:dms:us-east-1:388460667113:rep:JPJLT3VN3X75TXNK22G27HNKWIE7KE2X74BRAIQ",
    "replication_instance_class": "dms.t2.micro",
    "replication_instance_identifier": "turbottest17690",
    "title": "turbottest17690"
  }
]
✔ PASSED

Running SQL query: test-notfound-query.sql
null
✔ PASSED

Running SQL query: test-turbot-query.sql
[
  {
    "akas": [
      "arn:aws:dms:us-east-1:388460667113:rep:JPJLT3VN3X75TXNK22G27HNKWIE7KE2X74BRAIQ"
    ],
    "tags": {
      "foo": "bar"
    },
    "title": "turbottest17690"
  }
]
✔ PASSED

POSTTEST: tests/aws_dms_replication_instance

TEARDOWN: tests/aws_dms_replication_instance

SUMMARY:

1/1 passed.


```
</details>

# Example query results
<details>
  <summary>Results</summary>
  
### Basic info

```sql
select
  replication_instance_identifier,
  engine_version,
  instance_create_time,
  kms_key_id,
  publicly_accessible,
  replication_instance_arn,
  region
from
  aws_dms_replication_instance;
```
```
+---------------------------------+----------------+----------------------+-----------------------------------------------------------------------------+---------------------+--------------------------------------------------------------------------------+-----------+
| replication_instance_identifier | engine_version | instance_create_time | kms_key_id                                                                  | publicly_accessible | replication_instance_arn                                                       | region    |
+---------------------------------+----------------+----------------------+-----------------------------------------------------------------------------+---------------------+--------------------------------------------------------------------------------+-----------+
| test-steampipe1                 | 3.4.4          | 2021-04-29 12:21:36  | arn:aws:kms:us-east-1:388460667113:key/835c8d9c-5811-4219-be5a-ee4262d2aa47 | false               | arn:aws:dms:us-east-1:388460667113:rep:test                                    | us-east-1 |
| test-steampipe                  | 3.4.4          | 2021-04-28 10:30:12  | arn:aws:kms:us-east-1:388460667113:key/835c8d9c-5811-4219-be5a-ee4262d2aa47 | true                | arn:aws:dms:us-east-1:388460667113:rep:6PCGSTOTNMYORYRL4FEBLGYKS5EU6636F2ZRZPY | us-east-1 |
+---------------------------------+----------------+----------------------+-----------------------------------------------------------------------------+---------------------+--------------------------------------------------------------------------------+-----------+

```

### List replication instances having Auto Minor Version Upgrade feature disabled

```sql
select
  replication_instance_identifier,
  replication_instance_arn,
  engine_version,
  instance_create_time,
  auto_minor_version_upgrade,
  region
from
  aws_dms_replication_instance
where
  not auto_minor_version_upgrade;
```
```
+---------------------------------+--------------------------------------------------------------------------------+----------------+----------------------+----------------------------+-----------+
| replication_instance_identifier | replication_instance_arn                                                       | engine_version | instance_create_time | auto_minor_version_upgrade | region    |
+---------------------------------+--------------------------------------------------------------------------------+----------------+----------------------+----------------------------+-----------+
| test-steampipe                  | arn:aws:dms:us-east-1:388460667113:rep:6PCGSTOTNMYORYRL4FEBLGYKS5EU6636F2ZRZPY | 3.4.4          | 2021-04-28 10:30:12  | false                      | us-east-1 |
+---------------------------------+--------------------------------------------------------------------------------+----------------+----------------------+----------------------------+-----------+

```

### List replication instances provisioned with undesired (for example, dms.r5.16xlarge and dms.r5.24xlarge are not desired) instance class

```sql
select
  replication_instance_identifier,
  replication_instance_arn,
  engine_version,
  instance_create_time,
  replication_instance_class,
  region
from
  aws_dms_replication_instance
where
  replication_instance_class not in ('dms.r5.16xlarge', 'dms.r5.24xlarge');
```
```
+---------------------------------+--------------------------------------------------------------------------------+----------------+----------------------+----------------------------+-----------+
| replication_instance_identifier | replication_instance_arn                                                       | engine_version | instance_create_time | replication_instance_class | region    |
+---------------------------------+--------------------------------------------------------------------------------+----------------+----------------------+----------------------------+-----------+
| test-steampipe                  | arn:aws:dms:us-east-1:388460667113:rep:6PCGSTOTNMYORYRL4FEBLGYKS5EU6636F2ZRZPY | 3.4.4          | 2021-04-28 10:30:12  | dms.t3.micro               | us-east-1 |
| test-steampipe1                 | arn:aws:dms:us-east-1:388460667113:rep:test                                    | 3.4.4          | 2021-04-29 12:21:36  | dms.t3.micro               | us-east-1 |
+---------------------------------+--------------------------------------------------------------------------------+----------------+----------------------+----------------------------+-----------+

```

### List replication instances that are publicly accessible

```sql
select
  replication_instance_identifier,
  replication_instance_arn,
  publicly_accessible,
  region
from
  aws_dms_replication_instance
where
  publicly_accessible;
```
```
+---------------------------------+--------------------------------------------------------------------------------+---------------------+-----------+
| replication_instance_identifier | replication_instance_arn                                                       | publicly_accessible | region    |
+---------------------------------+--------------------------------------------------------------------------------+---------------------+-----------+
| test-steampipe                  | arn:aws:dms:us-east-1:388460667113:rep:6PCGSTOTNMYORYRL4FEBLGYKS5EU6636F2ZRZPY | true                | us-east-1 |
+---------------------------------+--------------------------------------------------------------------------------+---------------------+-----------+

```

### List replication instances not using Multi-AZ deployment configurations

```sql
select
  replication_instance_identifier,
  replication_instance_arn,
  publicly_accessible,
  multi_az,
  region
from
  aws_dms_replication_instance
where
  not multi_az;
```
```
+---------------------------------+--------------------------------------------------------------------------------+---------------------+----------+-----------+
| replication_instance_identifier | replication_instance_arn                                                       | publicly_accessible | multi_az | region    |
+---------------------------------+--------------------------------------------------------------------------------+---------------------+----------+-----------+
| test-steampipe                  | arn:aws:dms:us-east-1:388460667113:rep:6PCGSTOTNMYORYRL4FEBLGYKS5EU6636F2ZRZPY | true                | false    | us-east-1 |
+---------------------------------+--------------------------------------------------------------------------------+---------------------+----------+-----------+

```
</details>

### CODE REVIEW

File: aws/service.go [hunk 1]
```diff
@@ -290,6 +291,27 @@ func DaxService(ctx context.Context, d *plugin.QueryData, region string) (*dax.D
 	return svc, nil
 }
 
+// DatabaseMigrationService returns the service connection for AWS Database Migration service
+func DatabaseMigrationService(ctx context.Context, d *plugin.QueryData, region string) (*databasemigrationservice.DatabaseMigrationService, error) {
+	if region == "" {
+		return nil, fmt.Errorf("region must be passed DmsService")
```
Comment (Subhajit97) [inline aws/service.go:297]:
```suggestion
		return nil, fmt.Errorf("region must be passed DatabaseMigrationService")
```

File: aws/table_aws_dms_replication_instance.go [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "vpc_security_groups",
+				Description: "The VPC security group for the instance.",
+				Type:        proto.ColumnType_JSON,
+			},
+			{
+				Name:        "tags_src",
+				Description: "A list of tags currently associated with the DMS Replication Instance.",
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:148]:
```suggestion
				Description: "A list of tags currently associated with the replication instance.",
```

File: aws/table_aws_dms_replication_instance.go [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+				Type:        proto.ColumnType_JSON,
+				Hydrate:     getDmsReplicationInstanceTags,
+				Transform:   transform.FromField("TagList"),
+			},
+
+			// Standard columns
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:154]:
```suggestion
			// Steampipe standard columns
```

File: aws/table_aws_dms_replication_instance.go [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+				Transform:   transform.FromField("ReplicationInstanceArn").Transform(arnToAkas),
+			},
+		}),
+	}
+}
+
+//// LIST FUNCTION
+
+func listDmsReplicationInstances(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
+
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:181]:
```suggestion
```

File: aws/service.go [hunk 5]
```diff
@@ -290,6 +291,27 @@ func DaxService(ctx context.Context, d *plugin.QueryData, region string) (*dax.D
...
+	}
+	// have we already created and cached the service?
+	serviceCacheKey := fmt.Sprintf("dms-%s", region)
```
Comment (Subhajit97) [inline aws/service.go:300]:
```suggestion
	serviceCacheKey := fmt.Sprintf("databasemigrationservice-%s", region)
```

File: aws/table_aws_dms_replication_instance.go [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+			}
+			return !isLast
+		},
+	)
+	return nil, err
+}
+
+//// HYDRATE FUNCTIONS
+
+func getDmsReplicationInstance(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:213]:
```suggestion
func getDmsReplicationInstance(ctx context.Context, d *plugin.QueryData, _ *plugin.HydrateData) (interface{}, error) {
```

File: aws/table_aws_dms_replication_instance.go [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+}
+
+func getDmsReplicationInstanceTags(ctx context.Context, d *plugin.QueryData, h *plugin.HydrateData) (interface{}, error) {
+	plugin.Logger(ctx).Trace("getDmsReplicationInstanceTags")
+	var region string
+	matrixRegion := plugin.GetMatrixItem(ctx)[matrixKeyRegion]
+	if matrixRegion != nil {
+		region = matrixRegion.(string)
+	}
+	replicationInstanceArn := *h.Item.(*databasemigrationservice.ReplicationInstance).ReplicationInstanceArn
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:257]:
```suggestion
	replicationInstanceArn := h.Item.(*databasemigrationservice.ReplicationInstance).ReplicationInstanceArn
```

File: aws/table_aws_dms_replication_instance.go [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+
+	// Create service
+	svc, err := DatabaseMigrationService(ctx, d, region)
+	if err != nil {
+		return nil, err
+	}
+
+	params := &databasemigrationservice.ListTagsForResourceInput{
+		ResourceArn: &replicationInstanceArn,
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:266]:
```suggestion
		ResourceArn: replicationInstanceArn,
```

File: aws/table_aws_dms_replication_instance.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+	return replicationInstanceTags, nil
+}
+
+//// TRANSFORM FUNCTIONS
+
+func dmsReplicationInstanceTagListToTagsMap(_ context.Context, d *transform.TransformData) (interface{}, error) {
+	data := d.HydrateItem.(*databasemigrationservice.ListTagsForResourceOutput)
+	if data.TagList == nil {
+		return nil, nil
+	}
```
Comment (Subhajit97) [inline aws/table_aws_dms_replication_instance.go:283]: Do we need this block, as the function already returns `nil` at the end?

**Review by Subhajit97 — CHANGES_REQUESTED** [review]
@sayan133 There are few changes required. Please see comments, thanks!

**Review by Subhajit97 — APPROVED** [review]
LGTM

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**